File: src/llmeC.c
/*******************************************************************/
/* Computation of the log-likelihood and marginal posterior of size*/
/*******************************************************************/

#include "llmeC.h"
#include <R.h>
#include <Rmath.h>
#include <math.h>

void gllcmpmeC (double *v,
            int *n, 
            double *srd, 
            double *numrec, 
            double *rectime,
	    int *maxcoupons,
            int *K,
            double *cmpscale, 
            double *llik, 
	    int *verbose
			 ) {
  int i, iu, ni, Ki, maxc;
  double dmean,beta0,beta1,lnsd,scale,opt;
  int give_log0=0;
  double temp, loglik, sint, u;
  double zero=0.0;

  Ki=(*K);
  ni=(*n);
  maxc=(*maxcoupons);

  double *pdf = (double *) malloc(sizeof(double) * (Ki+1));
  double *rtprob = (double *) malloc(sizeof(double) * ni);

  dmean=(exp(v[0])+1.0)/exp(v[0]);
  beta0=v[1];
  beta1=v[2];
  lnsd=exp(v[3]);
  if((*cmpscale)<0.0){
    scale=exp(v[4]);
    opt=exp(v[5]);
  }else{
    scale=(*cmpscale);
    opt=exp(v[4]);
  }
  for (i=0; i<ni; i++){
    temp = beta0 + beta1*rectime[i];
    rtprob[i] = exp(temp)/(1.0+exp(temp));
//  Rprintf("%f %f %f %f\n",beta0,beta1,rtprob[i],temp);
  }
//  for (iu=0; iu<Ki; iu++){
//    pdf[iu]=exp((iu+1.0)*log(dmean)-scale*lgamma(iu+2.0));
//    temp+=pdf[iu];
////  pdf[iu] = dnbinom_mu(u,scale,dmean,give_log0);
//  }
////dcmp (intu, dmean, scale, Ki, 0.000000001, give_log0, pdf)
//  for (iu=0; iu<Ki; iu++){
//    pdf[iu]/=temp;
//  }
  pdf[0]=0.0;
  pdf[1]=1.0;
  temp=1.0;
  for (iu=2; iu<=Ki; iu++){
    u=iu-1.0;
    loglik=exp(u*log(dmean)-scale*lgamma(u+1.0));
    pdf[iu]=loglik;
    temp+=loglik;
//Rprintf("u %d pdf %f\n",iu,pdf[iu]);
  }
  for (iu=(1+Ki); iu<=(2*Ki); iu++){
    u=iu-1.0;
    loglik=exp(u*log(dmean)-scale*lgamma(u+1.0));
    pdf[Ki]+=loglik;
//Rprintf("u %d pdf Ki %f\n",iu,pdf[Ki]);
    temp+=loglik;
  }
//Rprintf("%f %f %f %f %f\n",dmean,scale,beta0,beta1,temp);
  for (iu=1; iu<=Ki; iu++){
    pdf[iu]/=temp;
//Rprintf("u %d pdf %f\n",iu,pdf[iu]);
  }

  loglik=0.0;
  for (i=0; i<ni; i++){
   sint = 0.0; 
   for (iu=1; iu<=Ki; iu++){
    temp = pdf[iu];
    u = (double)iu;
    if((iu <= maxc)|(numrec[i]<maxc)){
     temp *= dbinom(numrec[i],u,rtprob[i],give_log0);
//  Rprintf("< %f %f %f\n",rtprob[i],u,dbinom(numrec[i],u,rtprob[i],give_log0));
    }else{
     temp *= 1.0-pbinom(maxc-1.0,u,rtprob[i],give_log0,give_log0);
//  Rprintf("< %f\n",1.0-pbinom(maxc-1.0,u,rtprob[i],give_log0,give_log0));
    }
    temp *= dnorm(log(srd[i])-log(u)+log(opt),zero,lnsd,give_log0);
//  Rprintf("dn %f\n", dnorm(log(srd[i])-log(u)+log(opt),zero,lnsd,give_log0));
    sint += temp; 
   }
   loglik+=log(sint);
  }
  (*llik)=loglik;
  free(rtprob);
  free(pdf);
}

void gcmpmepdfC (double *v,
            int *n, 
            double *srd, 
            double *numrec, 
            double *rectime,
	    int *maxcoupons,
            int *K,
            double *cmpscale, 
            double *dpdf, 
	    int *verbose
			 ) {
  int i, iu, ni, Ki, maxc;
  double dmean,beta0,beta1,lnsd,scale,opt;
  int give_log0=0;
  double temp, sint, u;
  double zero=0.0;

  Ki=(*K);
  ni=(*n);
  maxc=(*maxcoupons);

  double *pdf = (double *) malloc(sizeof(double) * (Ki+1));
//int *intu = (double *) malloc(sizeof(int) * Ki);
  double *rtprob = (double *) malloc(sizeof(double) * ni);

  dmean=v[0];
  beta0=v[1];
  beta1=v[2];
  lnsd=v[3];
  if((*cmpscale)<0.0){
    scale=v[4];
    opt=v[5];
  }else{
    scale=(*cmpscale);
    opt=v[4];
  }
  for (i=0; i<ni; i++){
    temp = beta0 + beta1*rectime[i];
    rtprob[i] = exp(temp)/(1.0+exp(temp));
  }

  temp=1.0;
  pdf[0]=0.0;
  pdf[1]=1.0;
  for (iu=2; iu<=Ki; iu++){
    u=iu-1.0;
    pdf[iu]=exp(u*log(dmean)-scale*lgamma(u+1.0));
    temp+=pdf[iu];
  }
  for (iu=(1+Ki); iu<=(2*Ki); iu++){
//  u=(double)iu;
    u=iu-1.0;
    sint=exp(u*log(dmean)-scale*lgamma(u+1.0));
    pdf[Ki]+=sint;
    temp+=sint;
  }
  for (iu=1; iu<=Ki; iu++){
    pdf[iu]/=temp;
  }

  for (i=0; i<ni; i++){
   sint = 0.0; 
   for (iu=1; iu<=Ki; iu++){
    temp = pdf[iu];
    u = (double)iu;
    if((iu <= maxc)|(numrec[i]<maxc)){
     temp *= dbinom(numrec[i],u,rtprob[i],give_log0);
//  Rprintf("< %f %f %f\n",rtprob[i],u,dbinom(numrec[i],u,rtprob[i],give_log0));
    }else{
     temp *= 1.0-pbinom(maxc-1.0,u,rtprob[i],give_log0,give_log0);
//  Rprintf("< %f\n",1.0-pbinom(maxc-1.0,u,rtprob[i],give_log0,give_log0));
    }
    if(srd[i]>0.0){
     temp *= dnorm(log(srd[i])-log(u)+log(opt),zero,lnsd,give_log0);
    }
//  Rprintf("dn %f\n", dnorm(log(srd[i])-log(u)+log(opt),zero,lnsd,give_log0));
    sint += temp; 
    dpdf[i*Ki+iu-1] = temp; 
   }
   for (iu=1; iu<=Ki; iu++){
    dpdf[i*Ki+iu-1] /= sint; 
   }
  }
  free(rtprob);
  free(pdf);
}

void gllnbmeC (double *v,
            int *n, 
            double *srd, 
            double *numrec, 
            double *rectime,
	    int *maxcoupons,
            int *K,
            double *nbscale, 
            double *llik, 
	    int *verbose
			 ) {
  int i, iu, ni, Ki, maxc;
  double dmean,beta0,beta1,lnsd,scale,opt;
  int give_log0=0;
  double temp, loglik, sint, u;
  double zero=0.0;

  Ki=(*K);
  ni=(*n);
  maxc=(*maxcoupons);

  double *pdf = (double *) malloc(sizeof(double) * (Ki+1));
  double *rtprob = (double *) malloc(sizeof(double) * ni);

  dmean=(exp(v[0])+1.0)/exp(v[0]);
  beta0=v[1];
  beta1=v[2];
  lnsd=exp(v[3]);
  if((*nbscale)<0.0){
    scale=dmean / exp(v[4]);
    opt=exp(v[5]);
  }else{
    scale=(*nbscale);
    opt=exp(v[4]);
  }
  for (i=0; i<ni; i++){
    temp = beta0 + beta1*rectime[i];
    rtprob[i] = exp(temp)/(1.0+exp(temp));
//  Rprintf("%f %f %f %f\n",beta0,beta1,rtprob[i],temp);
  }
  pdf[0]=0.0;
  temp=0.0;
  for (iu=1; iu<=Ki; iu++){
    u=((double)iu)-1.0;
    loglik = dnbinom_mu(u,scale,dmean,give_log0);
    temp+=loglik;
    pdf[iu] = loglik;
  }
  pdf[Ki]+=1.0-temp;
//pdf[Ki]+=1.0-temp-dnbinom_mu(0.0,scale,dmean,give_log0);
//temp=1.0-dnbinom_mu(0.0,scale,dmean,give_log0);
//for (iu=1; iu<=Ki; iu++){
//  pdf[iu]/=temp;
//}

  loglik=0.0;
  for (i=0; i<ni; i++){
   sint = 0.0; 
   for (iu=1; iu<=Ki; iu++){
    temp = pdf[iu];
    u = (double)iu;
    if((iu <= maxc)|(numrec[i]<maxc)){
     temp *= dbinom(numrec[i],u,rtprob[i],give_log0);
//  Rprintf("< %f %f %f\n",rtprob[i],u,dbinom(numrec[i],u,rtprob[i],give_log0));
    }else{
     temp *= 1.0-pbinom(maxc-1.0,u,rtprob[i],give_log0,give_log0);
//  Rprintf("< %f\n",1.0-pbinom(maxc-1.0,u,rtprob[i],give_log0,give_log0));
    }
    temp *= dnorm(log(srd[i])-log(u)+log(opt),zero,lnsd,give_log0);
//  Rprintf("dn %f\n", dnorm(log(srd[i])-log(u)+log(opt),zero,lnsd,give_log0));
    sint += temp; 
   }
   loglik+=log(sint);
  }
  (*llik)=loglik;
  free(rtprob);
  free(pdf);
}

void gnbmepdfC (double *v,
            int *n, 
            double *srd, 
            double *numrec, 
            double *rectime,
	    int *maxcoupons,
            int *K,
            double *nbscale, 
            double *dpdf, 
	    int *verbose
			 ) {
  int i, iu, ni, Ki, maxc;
  double dmean,beta0,beta1,lnsd,scale,opt;
  int give_log0=0;
  double temp, sint, u;
  double zero=0.0;

  Ki=(*K);
  ni=(*n);
  maxc=(*maxcoupons);

  double *pdf = (double *) malloc(sizeof(double) * (Ki+1));
  double *rtprob = (double *) malloc(sizeof(double) * ni);

  dmean=v[0];
  beta0=v[1];
  beta1=v[2];
  lnsd=v[3];
  if((*nbscale)<zero){
    scale=dmean / (v[4]-1.0);
    opt=v[5];
  }else{
    scale=(*nbscale);
    opt=v[4];
  }
  for (i=0; i<ni; i++){
    temp = beta0 + beta1*rectime[i];
    rtprob[i] = exp(temp)/(1.0+exp(temp));
  }
  pdf[0]=zero;
  temp=zero;
  for (iu=1; iu<=Ki; iu++){
    u=((double)iu)-1.0;
    sint = dnbinom_mu(u,scale,dmean,give_log0);
    temp+=sint;
    pdf[iu] = sint;
  }
  pdf[Ki]+=1.0-temp;
//pdf[Ki]+=1.0-temp-dnbinom_mu(0.0,scale,dmean,give_log0);
//temp=1.0-dnbinom_mu(0.0,scale,dmean,give_log0);
//for (iu=1; iu<=Ki; iu++){
//  pdf[iu]/=temp;
//}

  for (i=0; i<ni; i++){
   sint = zero; 
   for (iu=1; iu<=Ki; iu++){
    temp = pdf[iu];
    u = (double)iu;
    if((iu <= maxc)|(numrec[i]<maxc)){
     temp *= dbinom(numrec[i],u,rtprob[i],give_log0);
//  Rprintf("< %f %f %f\n",rtprob[i],u,dbinom(numrec[i],u,rtprob[i],give_log0));
    }else{
     temp *= 1.0-pbinom(maxc-1.0,u,rtprob[i],give_log0,give_log0);
//  Rprintf("< %f\n",1.0-pbinom(maxc-1.0,u,rtprob[i],give_log0,give_log0));
    }
    if(srd[i]>zero){
     temp *= dnorm(log(srd[i])-log(u)+log(opt),zero,lnsd,give_log0);
    }
//  Rprintf("dn %f\n", dnorm(log(srd[i])-log(u)+log(opt),zero,lnsd,give_log0));
    sint += temp; 
    dpdf[i*Ki+iu-1] = temp; 
   }
   for (iu=1; iu<=Ki; iu++){
    dpdf[i*Ki+iu-1] /= sint; 
   }
  }
  free(rtprob);
  free(pdf);
}

File: R/llmeC.R
memle <-function(network.size,num.recruits,recruit.time,max.coupons=3,unit.scale=NULL,

          unit.model="nbinom",

          cutoff=0,cutabove=1000,

          guess=c(3,0,0,0.6,1.5,0),

          method="BFGS", hessian=TRUE, K=max(100,network.size), verbose=FALSE){

 logit <- function(p){log(p/(1-p))}

 if(is.null(unit.scale)){

   unit.scale <- -1

   vtrans <- function(l){c((exp(l[1])+1)/exp(l[1]),l[2],l[3],exp(l[4]),exp(l[5])+1,exp(l[6]))}

   ltrans <- function(v){c(-log(v[1]-1),v[2],v[3],log(v[4]),log(v[5]-1),log(v[6]))}

   gtrans <- function(l){c(-exp(-l[1]),1,1,exp(l[4]),exp(l[5]),exp(l[6]))}

 }else{

   unit.scale <- 1

   vtrans <- function(l){c((exp(l[1])+1)/exp(l[1]),l[2],l[3],exp(l[4]),exp(l[5]))}

   ltrans <- function(v){c(-log(v[1]-1),v[2],v[3],log(v[4]),log(v[5]))}

   gtrans <- function(l){c(-exp(-l[1]),1,1,exp(l[4]),exp(l[5]))}

 }

 n <- length(network.size)

#

 llme <- switch(unit.model,

  "cmp"=llcmpme, "nbinom"=llnbme)

 if(sum(network.size>=cutoff & network.size <= cutabove) > 0){

  aaa <- optim(par=ltrans(guess),fn=llme,

   method=method,

   hessian=hessian,control=list(fnscale=-10),

   n=n,

   recruit.time=recruit.time,

   num.recruits=num.recruits,network.size=network.size,

   max.coupons=max.coupons,K=K,

   unit.scale=unit.scale, verbose=verbose)

   if(unit.scale<0){

    if(unit.model=="cmp"){

     names(aaa$par) <- c("CMP mean","Recruitment Odds","Recruitment Odds Time","Error log-s.d.", "CMP skew","Optimism")

    }else{

     names(aaa$par) <- c("Neg.Bin. mean","Recruitment Odds","Recruitment Odds Time","Error log-s.d.", "Neg.Bin scale","Optimism")

    }

   }else{

    if(unit.scale<0){

     names(aaa$par) <- c("CMP mean","Recruitment Odds","Recruitment Odds Time","Error log-s.d.","Optimism")

    }else{

     names(aaa$par) <- c("Neg.Bin. mean","Recruitment Odds","Recruitment Odds Time","Error log-s.d.","Optimism")

    }

   }

   v=vtrans(aaa$par)

# if(is.psd(-aaa$hessian)){

  if(TRUE){

   g=gtrans(aaa$par)

   asycov = diag(g) %*% robust.inverse(-aaa$hessian) %*% diag(g)

   dimnames(asycov) <- list(names(aaa$par),names(aaa$par))

   asyse <- sqrt(diag(asycov))

   asycor <- diag(1/asyse) %*% asycov %*% diag(1/asyse)

   dimnames(asycor) <- dimnames(asycov)

   names(asyse) <- names(aaa$par)

   ccc <- list(coef=v,iterations=as.numeric(aaa$counts[1]),

               covar=asycov,se=asyse,asycor=asycor,

               df=length(network.size),loglik=aaa$value,

               loglik.null=-length(network.size)*(log(max(network.size))+log(max.coupons+1)))

  }else{

   ccc <- list(theta=vtrans(aaa$par))

  }

 }else{

  ccc <- list(theta=rep(NA,length=6))

 }

 class(ccc) <- "me"

 ccc

}

memle.nb <-function(network.size,num.recruits,recruit.time,max.coupons=3,nb.scale=NULL,cutoff=0,cutabove=1000,

          guess=c(3,0,0,0.6,1,0),

          method="BFGS", hessian=TRUE, K=max(100,network.size), verbose=FALSE){

 logit <- function(p){log(p/(1-p))}

 if(is.null(nb.scale)){

   nb.scale <- -1

   vtrans <- function(l){c((exp(l[1])+1)/exp(l[1]),l[2],l[3],exp(l[4]),exp(l[5])+1,exp(l[6]))}

   ltrans <- function(v){c(-log(v[1]-1),v[2],v[3],log(v[4]),log(v[5]-1),log(v[6]))}

   gtrans <- function(l){c(-exp(-l[1]),1,1,exp(l[4]),exp(l[5]),exp(l[6]))}

 }else{

   nb.scale <- 1

   vtrans <- function(l){c((exp(l[1])+1)/exp(l[1]),l[2],l[3],exp(l[4]),exp(l[5]))}

   ltrans <- function(v){c(-log(v[1]-1),v[2],v[3],log(v[4]),log(v[5]))}

   gtrans <- function(l){c(-exp(-l[1]),1,1,exp(l[4]),exp(l[5]))}

 }

 n <- length(network.size)

#

 if(sum(network.size>=cutoff & network.size <= cutabove) > 0){

  aaa <- optim(par=ltrans(guess),fn=llnbme,

   method=method,

   hessian=hessian,control=list(fnscale=-10),

   n=n,

   recruit.time=recruit.time,

   num.recruits=num.recruits,network.size=network.size,

   max.coupons=max.coupons,K=K,

   nb.scale=nb.scale, verbose=verbose)

   if(nb.scale<0){

     names(aaa$par) <- c("Neg.Bin. mean","Recruitment Odds","Recruitment Odds Time","Error log-s.d.", "Neg.Bin scale","Optimism")

   }else{

     names(aaa$par) <- c("Neg.Bin. mean","Recruitment Odds","Recruitment Odds Time","Error log-s.d.","Optimism")

   }

   v=vtrans(aaa$par)

# if(is.psd(-aaa$hessian)){

  if(TRUE){

   g=gtrans(aaa$par)

   asycov = diag(g) %*% robust.inverse(-aaa$hessian) %*% diag(g)

   dimnames(asycov) <- list(names(aaa$par),names(aaa$par))

   asyse <- sqrt(diag(asycov))

   asycor <- diag(1/asyse) %*% asycov %*% diag(1/asyse)

   dimnames(asycor) <- dimnames(asycov)

   names(asyse) <- names(aaa$par)

   ccc <- list(coef=v,iterations=as.numeric(aaa$counts[1]),

               covar=asycov,se=asyse,asycor=asycor,

               df=length(network.size),loglik=aaa$value,

               loglik.null=-length(network.size)*(log(max(network.size))+log(max.coupons+1)))

  }else{

   ccc <- list(theta=vtrans(aaa$par))

  }

 }else{

  ccc <- list(theta=rep(NA,length=6))

 }

 class(ccc) <- "me"

 ccc

}

llnbme <- function(v,n,network.size,num.recruits,recruit.time,

                  max.coupons,K=100,unit.scale=-1,verbose=FALSE){

 Cret <- .C("gllnbmeC",

           v=as.double(v),

           n=as.integer(n),

           srd=as.double(network.size),

           numrec=as.double(num.recruits),

           rectime=as.double(recruit.time),

           maxcoupons=as.integer(max.coupons),

           K=as.integer(K),

           nb.scale=as.double(unit.scale),

           llik=as.double(0),

           verbose=as.integer(verbose), PACKAGE="RDS")

 out=Cret$llik

 if(verbose){cat(sprintf("llik=%f\n", out))}

 if(is.infinite(out)){out <- -10^10}

 if(is.na(out)){out <- -10^10}

 out

}



dnbmepdf <- function(v,network.size,num.recruits,recruit.time,max.coupons=3,K=100,nb.scale=NULL,verbose=FALSE){

 if(is.null(nb.scale)){

   nb.scale <- -1

 }else{

   nb.scale <- 1

 }

 n <- length(network.size)

 network.size[is.na(network.size)] <- -1

 Cret <- .C("gnbmepdfC",

           v=as.double(v),

           n=as.integer(n),

           srd=as.double(network.size),

           numrec=as.double(num.recruits),

           rectime=as.double(recruit.time),

           maxcoupons=as.integer(max.coupons),

           K=as.integer(K),

           nb.scale=as.double(nb.scale),

           pdf=double(n*K),

           verbose=as.integer(verbose), PACKAGE="RDS")



  return(matrix(Cret$pdf,ncol=n,nrow=K))

}

llcmpme <- function(v,n,network.size,num.recruits,recruit.time,

                  max.coupons,K=100,unit.scale=-1,verbose=FALSE){

 Cret <- .C("gllcmpmeC",

           v=as.double(v),

           n=as.integer(n),

           srd=as.double(network.size),

           numrec=as.double(num.recruits),

           rectime=as.double(recruit.time),

           maxcoupons=as.integer(max.coupons),

           K=as.integer(K),

           cmp.scale=as.double(unit.scale),

           llik=as.double(0),

           verbose=as.integer(verbose), PACKAGE="RDS")

 out=Cret$llik

 if(verbose){cat(sprintf("llik=%f\n", out))}

 if(is.infinite(out)){out <- -10^10}

 if(is.na(out)){out <- -10^10}

 out

}



dmepdf <- function(v,network.size,num.recruits,recruit.time,max.coupons=3,K=100,unit.scale=NULL,unit.model="nbinom",verbose=FALSE){

 if(is.null(unit.scale)){

   unit.scale <- -1

 }else{

   unit.scale <- 1

 }

 n <- length(network.size)

 network.size[is.na(network.size)] <- -1

 if(unit.model=="cmp"){

   Cret <- .C("gcmpmepdfC",

             v=as.double(v),

             n=as.integer(n),

             srd=as.double(network.size),

             numrec=as.double(num.recruits),

             rectime=as.double(recruit.time),

             maxcoupons=as.integer(max.coupons),

             K=as.integer(K),

             cmp.scale=as.double(unit.scale),

             pdf=double(n*K),

             verbose=as.integer(verbose), PACKAGE="RDS")

 }else{

   Cret <- .C("gnbmepdfC",

             v=as.double(v),

             n=as.integer(n),

             srd=as.double(network.size),

             numrec=as.double(num.recruits),

             rectime=as.double(recruit.time),

             maxcoupons=as.integer(max.coupons),

             K=as.integer(K),

             nb.scale=as.double(unit.scale),

             pdf=double(n*K),

             verbose=as.integer(verbose), PACKAGE="RDS")

 }



  return(matrix(Cret$pdf,ncol=n,nrow=K))

}

